### PR TITLE
Add WASM SEGV handler and test.  Create and use a Makefile.base for WASM c++ compiles.

### PR DESCRIPTION
**Pull request by jplevyak**
Create and use a Makefile.base for WASM c++ compiles.

For an explanation of how to fill out the fields, please see the relevant section 
in [PULL_REQUESTS.md](./PULL_REQUESTS.md)

*Description*:
*Risk Level*:
*Testing*:
*Docs Changes*:
*Release Notes*:
[Optional Fixes #Issue]
[Optional *Deprecated*:]

### CODE REVIEW

File: api/wasm/cpp/Makefile.base [hunk 1]
```diff
@@ -0,0 +1,8 @@
+API:=$(shell git rev-parse --show-toplevel)/api/wasm/cpp
```
Comment (silentdai) [inline api/wasm/cpp/Makefile.base]: Seems only reasonable if user use git.  I am fine with current phase

File: test/extensions/wasm/test_data/segv.cc [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// NOLINT(namespace-envoy)
+#include <string>
+
+#include "proxy_wasm_intrinsics.h"
+
+static int *badptr = nullptr;
+
+extern "C" EMSCRIPTEN_KEEPALIVE void proxy_onStart() {
+  logError("before badptr");
+  *badptr = 1;
```
Comment (silentdai) [inline test/extensions/wasm/test_data/segv.cc]: how is this pointer initialized?

File: test/extensions/wasm/wasm_test.cc [hunk 3]
```diff
@@ -73,6 +73,25 @@ TEST(WasmTest, BadSignature) {
...
+TEST(WasmTest, Segv) {
+  Stats::IsolatedStoreImpl stats_store;
+  Api::ApiPtr api = Api::createApiForTest(stats_store);
+  Upstream::MockClusterManager cluster_manager;
+  Event::SimulatedTimeSystem time_system;
+  Event::DispatcherImpl dispatcher(time_system, *api);
+  auto wasm = std::make_unique<Extensions::Common::Wasm::Wasm>("envoy.wasm.vm.wavm", "", "",
+                                                               cluster_manager, dispatcher);
+  EXPECT_NE(wasm, nullptr);
+  const auto code = TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(
```
Comment (silentdai) [inline test/extensions/wasm/wasm_test.cc]: This is fancy!

File: test/extensions/wasm/wasm_test.cc [hunk 4]
```diff
@@ -73,6 +73,25 @@ TEST(WasmTest, BadSignature) {
                             "Bad function signature for: _proxy_onConfigure");
 }
 
+TEST(WasmTest, Segv) {
+  Stats::IsolatedStoreImpl stats_store;
+  Api::ApiPtr api = Api::createApiForTest(stats_store);
+  Upstream::MockClusterManager cluster_manager;
+  Event::SimulatedTimeSystem time_system;
+  Event::DispatcherImpl dispatcher(time_system, *api);
+  auto wasm = std::make_unique<Extensions::Common::Wasm::Wasm>("envoy.wasm.vm.wavm", "", "",
```
Comment (silentdai) [inline test/extensions/wasm/wasm_test.cc]: make_shared?

File: source/extensions/common/wasm/wasm.cc [hunk 5]
```diff
@@ -441,6 +441,10 @@ uint32_t httpCallHandler(void* raw_context, uint32_t uri_ptr, uint32_t uri_size,
 
 uint32_t getTotalMemoryHandler(void*) { return 0x7FFFFFFF; }
 uint32_t _emscripten_get_heap_sizeHandler(void*) { return 0x7FFFFFFF; }
+void _llvm_trapHandler(void*) {
+  std::cerr << "llvm_trapHandler";
```
Comment (silentdai) [inline source/extensions/common/wasm/wasm.cc]: cerr? It's the first time I see the usage of sderr in the code but you may be right

File: api/wasm/cpp/Makefile.base (hunk 1 above)
Comment (jplevyak) [reply api/wasm/cpp/Makefile.base]: I imagine that users will have to changes this file or link to the api in their project.

File: test/extensions/wasm/test_data/segv.cc (hunk 2 above)
Comment (jplevyak) [reply test/extensions/wasm/test_data/segv.cc]: It is a static initialized to nullptr.

File: source/extensions/common/wasm/wasm.cc (hunk 5 above)
Comment (jplevyak) [reply source/extensions/common/wasm/wasm.cc]: Debugging code, removing...

File: test/extensions/wasm/wasm_test.cc (hunk 4 above)
Comment (jplevyak) [reply test/extensions/wasm/wasm_test.cc]: done.

**Comment by jplevyak** [comment]
This changed converts the crash into an Exception which we can then handle.  When llvm finds an situation where an SEGV is guaranteed, but which it can detect, it calls llvm_trap.  This just converts that call into an Exception.

**Comment by jplevyak** [comment]
Created https://github.com/istio/envoy/pull/26.  I will create the Makefile.base PR when the fix for -O3.

**Pull request closed without merging**